File: ios/Stopwatch/core/utils/ObservableState.swift
import Foundation

protocol StateObserver<T>: Identifiable {
    associatedtype T
    
    func onStateChange(_ state: T)
}

class ObservableState<S: StateObserver> {
    private(set) var state: S.T
    
    private var observers: Dictionary<S.ID, S> = Dictionary()
    
    init(_ state: S.T) {
        self.state = state
    }

    func addObserver(_ observer: S) {
        observers[observer.id] = observer
    }
    
    func removeObserver(_ observer: S) {
        observers[observer.id] = nil
    }
    
    func updateState(_ state: S.T) {
        self.state = state

        for (_, observer) in observers {
            observer.onStateChange(self.state)
        }
    }
}

File: android/app/src/main/java/com/rafaeltmbr/stopwatch/core/utils/ObservableState.kt
package com.rafaeltmbr.stopwatch.core.utils

interface StateObserver<T> {
    fun onStateChange(state: T)
}

class ObservableState<T>(private var _state: T) {
    private val stateObservers: MutableSet<StateObserver<T>> = mutableSetOf()

    val state: T
        get() = this._state

    fun addObserver(stateObserver: StateObserver<T>) {
        stateObservers.add(stateObserver)
    }

    fun removeObserver(stateObserver: StateObserver<T>) {
        stateObservers.remove(stateObserver)
    }

    fun updateState(state: T) {
        this._state = state

        for (observer in stateObservers) {
            observer.onStateChange(this._state)
        }
    }
}